lua/cfc_random_spawn/module/sv_player_spawning.lua

```diff
@@ -217,20 +297,28 @@ local function getPopularCenter( plys )
     return bestCenter
 end
 
+local function updatePopularCenter( measurablePlayers )
+    if CFCRandomSpawn.doDynamicCenters then
+        mostPopularCenter = getDynamicPvpCenter( measurablePlayers )
+    else
+        mostPopularCenter = getPopularCenter( measurablePlayers )
+    end
+    CFCRandomSpawn.mostPopularCenter = mostPopularCenter
+
+    CENTER_CUTOFF = mostPopularCenter.overrideCutoff or DEFAULT_CENTER_CUTOFF
+    CENTER_CUTOFF_SQR = mostPopularCenter.overrideCutoffSqr or DEFAULT_CENTER_CUTOFF_SQR
+end
+
 function CFCRandomSpawn.getOptimalSpawnPos()
     local measurablePlayers = getMeasurablePlayers()
     local allLivingPlys = getLivingPlayers()
 
-    if CENTER_UPDATE_ON_RESPAWN then
-        mostPopularCenter = getPopularCenter( measurablePlayers ) or pvpCenters[1]
-        CFCRandomSpawn.mostPopularCenter = mostPopularCenter
-
-        CENTER_CUTOFF = mostPopularCenter.overrideCutoff or DEFAULT_CENTER_CUTOFF
-        CENTER_CUTOFF_SQR = mostPopularCenter.overrideCutoffSqr or DEFAULT_CENTER_CUTOFF_SQR
+    if CENTER_UPDATE_ON_RESPAWN or not CFCRandomSpawn.mostPopularCenter then
+        updatePopularCenter( measurablePlayers )
     end
 
     local freeSpawns = findFreeSpawnPoints( customSpawnsForMap, allLivingPlys )
-    local nearestSpawns = getNearestSpawns( getPlyAvg( measurablePlayers, mostPopularCenter.centerPos ), freeSpawns )
+    local nearestSpawns = getNearestSpawns( getPlyAvg( measurablePlayers, CFCRandomSpawn.mostPopularCenter.centerPos ), freeSpawns )
```

Comment:
This also isnt in the scope of this PR but it looks like every spawn. We are calculating if every spawn is free, then getting the nearest spawns, then sorting them. 

I think most spawns will usually be free. So depending on how many spawns exist It might be more efficient to instead get the nearest spawns without calculating if they are free, pick a random one, check if its free, then repeat until you get a free spawn.

Reply:
Then we could cache the nearest spawns table for a few seconds and avoid recalculating it every spawn.

Reply:
good changes but they're out of scope for now.
i just made getNearestSpawns run before findFreeSpawnPoints

Reply:
Nearest spawns table is also cached now.